spec/validator/markdown_validator_spec.rb

```diff
@@ -0,0 +1,56 @@
+require "rails_helper"
+
+class MarkdownModelWithHeadingsAllowed
+  include ActiveModel::Model
+  attr_accessor :markdown
+
+  validates :markdown, markdown: { allow_headings: true }
+end
+
+class MarkdownModelWithHeadingsDisallowed
+  include ActiveModel::Model
+  attr_accessor :markdown
+
+  validates :markdown, markdown: { allow_headings: false }
+end
+
+RSpec.describe MarkdownValidator do
+  let(:markdown_model) do
+    MarkdownModelWithHeadingsAllowed.new
+  end
+
+  it "is invalid if guidance markdown contains unsupported tags" do
+    markdown_model.markdown = "# Heading level 1"
+    expect(markdown_model).to be_invalid
+    expect(markdown_model.errors.map(&:type)).to include :unsupported_markdown_syntax
+  end
+
+  it "is valid if guidance markdown contains supported headings" do
+    markdown_model.markdown = "## Heading level 2"
+    expect(markdown_model).to be_valid
+  end
+
+  it "is invalid if guidance markdown is over 5000 characters" do
+    markdown_model.markdown = "A" * 5001
+    expect(markdown_model).to be_invalid
+    expect(markdown_model.errors.map(&:type)).to include :too_long
+  end
+
+  context "when headings are not allowed" do
+    let(:markdown_model) do
+      MarkdownModelWithHeadingsDisallowed.new
+    end
+
+    it "is invalid if guidance markdown contains level 2 headings" do
+      markdown_model.markdown = "## Heading level 2"
+      expect(markdown_model).to be_invalid
+      expect(markdown_model.errors.map(&:type)).to include :unsupported_markdown_syntax
+    end
+
+    it "is invalid if guidance markdown contains level 3 headings" do
+      markdown_model.markdown = "### Heading level 3"
+      expect(markdown_model).to be_invalid
+      expect(markdown_model.errors.map(&:type)).to include :unsupported_markdown_syntax
+    end
+  end
```

Comment:
a shared_example might be good here so we could reuse it for the what happens next